packages/server/next/app/components/ui/chat/custom/events/loader.ts

```diff
@@ -1,71 +1,190 @@
-import * as Babel from "@babel/standalone"; // Import Babel standalone for runtime transpilation
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { parse } from "@babel/parser";
+import * as Babel from "@babel/standalone";
+import traverse from "@babel/traverse";
 import { JSONValue } from "@llamaindex/chat-ui";
 import React, { FunctionComponent } from "react";
 import { getConfig } from "../../../lib/utils";
 import { ComponentDef } from "./types";
 
 export type SourceComponentDef = {
-  type: string; // eg. deep_research_event
-  code: string; // source code for component
-  filename: string; // eg. deep_research_event.tsx
+  type: string;
+  code: string;
+  filename: string;
 };
 
 export async function fetchComponentDefinitions(): Promise<ComponentDef[]> {
   const endpoint = getConfig("COMPONENTS_API");
   if (!endpoint) return [];
 
-  try {
-    const response = await fetch(endpoint);
-    const components = (await response.json()) as SourceComponentDef[];
-
-    // Only need to handle transpilation now
-    const transpiledComponents = await Promise.all(
-      components.map(async (comp) => ({
-        type: comp.type,
-        comp: await transpileCode(comp.code, comp.filename),
-      })),
-    );
+  const response = await fetch(endpoint);
+  const components = (await response.json()) as SourceComponentDef[];
 
-    return transpiledComponents.filter(
-      (comp): comp is ComponentDef => comp.comp !== null,
-    );
-  } catch (error) {
-    console.log("Error fetching dynamic components:", error);
-    return [];
-  }
+  // Only need to handle transpilation now
+  const transpiledComponents = await Promise.all(
+    components.map(async (comp) => ({
+      type: comp.type,
+      comp: await parseComponent(comp.code, comp.filename),
+    })),
+  );
+
+  return transpiledComponents.filter(
+    (comp): comp is ComponentDef => comp.comp !== null,
+  );
 }
 
-// convert TSX code to JS code using Babel
-export async function transpileCode(
+// create React component from code
+async function parseComponent(
   code: string,
   filename: string,
 ): Promise<FunctionComponent<{ events: JSONValue[] }> | null> {
```

Comment:
adding EventRenderComponent to types?
```suggestion
): Promise<EventRenderComponent | null> {
```